cpp/src/parquet/metadata_benchmark.cc

```diff
@@ -0,0 +1,152 @@
+    writer_properties_ =
+        prop_builder.version(ParquetVersion::PARQUET_2_6)->disable_dictionary()->build();
+  }
+
+  std::shared_ptr<BufferOutputStream> WriteFile() {
+    PARQUET_ASSIGN_OR_THROW(auto sink, BufferOutputStream::Create());
+
+    auto writer = ParquetFileWriter::Open(sink, schema_root_, writer_properties_);
+    std::vector<int32_t> int32_values(42, 1);
+    for (int rg = 0; rg < num_row_groups_; ++rg) {
+      auto row_group_writer = writer->AppendRowGroup();
+      for (int col = 0; col < num_columns_; ++col) {
+        auto col_writer = row_group_writer->NextColumn();
+        ARROW_CHECK_EQ(col_writer->type(), Type::INT32);
+        auto typed_col_writer = static_cast<Int32Writer*>(col_writer);
+        typed_col_writer->WriteBatch(
+            /*num_values=*/static_cast<int64_t>(int32_values.size()),
+            /*def_levels=*/nullptr, /*rep_levels=*/nullptr, int32_values.data());
+        typed_col_writer->Close();
+      }
+      row_group_writer->Close();
+    }
+    writer->Close();
+    return sink;
+  }
+
+  void ReadFile(std::shared_ptr<Buffer> contents) {
+    auto source = std::make_shared<BufferReader>(contents);
+    ReaderProperties props;
+    auto reader = ParquetFileReader::Open(source, props);
+    auto metadata = reader->metadata();
+    ARROW_CHECK_EQ(metadata->num_columns(), num_columns_);
+    ARROW_CHECK_EQ(metadata->num_row_groups(), num_row_groups_);
+    reader->Close();
+  }
+
+ private:
+  int num_columns_;
+  int num_row_groups_;
+  std::shared_ptr<GroupNode> schema_root_;
+  std::shared_ptr<WriterProperties> writer_properties_;
+};
+
+void WriteMetadataSetArgs(benchmark::internal::Benchmark* bench) {
+  bench->ArgNames({"num_columns", "num_row_groups"});
+
+  for (int num_columns : {1, 10, 100}) {
+    for (int num_row_groups : {1, 100, 1000}) {
+      bench->Args({num_columns, num_row_groups});
+    }
+  }
+  /* For larger num_columns, restrict num_row_groups to small values
+   * to avoid blowing up benchmark execution time.
+   */
+  for (int num_columns : {1000, 10000}) {
+    for (int num_row_groups : {1, 100}) {
+      bench->Args({num_columns, num_row_groups});
+    }
+  }
+}
+
+void ReadMetadataSetArgs(benchmark::internal::Benchmark* bench) {
+  WriteMetadataSetArgs(bench);
+}
+
+void WriteMetadata(benchmark::State& state) {
+  MetadataBenchmark benchmark(&state);
+
+  int64_t file_size = 0;
+  for (auto _ : state) {
+    auto sink = benchmark.WriteFile();
+    file_size = sink->Finish().ValueOrDie()->size();
+  }
+  state.SetItemsProcessed(state.iterations());
+  state.counters["file_size"] = static_cast<double>(file_size);
+}
+
+void ReadMetadata(benchmark::State& state) {
+  MetadataBenchmark benchmark(&state);
+  auto contents = benchmark.WriteFile()->Finish().ValueOrDie();
+
+  for (auto _ : state) {
+    benchmark.ReadFile(contents);
+  }
+  state.SetItemsProcessed(state.iterations());
+  state.counters["file_size"] = static_cast<double>(contents->size());
```

Comment:
add `metadata()->SerializeToString().size()` here?

Reply:
What additional information would it give?

Reply:
File has been write with different batches. The file_size might also changed if `encoding` changed. Instead of batch-size, shouldn't we focus more on file metadata thrift size?

Reply:
> File has been write with different batches with multiple columns.

Varying the number of columns is important here.

> The file_size might also changed if `encoding` changed.

Did you check the data? It's tiny (one INT32 value per column chunk). The entire file will consist of metadata overhead.

Reply:
> Did you check the data? It's tiny (one INT32 value per column chunk). The entire file will consist of metadata overhead.

I'll download one. I think we'd better get "concrete" size of footer, but we can choose current impl if you consist

( Besides, footer read by default would be 64KiB, if the footer is greater than 64KiB, a more around of read and parse might be run here)

Reply:
After checking, I've generate the case for 
* `ReadMetadata/num_columns:10000/num_row_groups:1`, the result is: `Footer offset: 2901570, size: 920550`. Footer occupies 900K and is 1/3 less than file size.

I think this might because some metadata has been "dual" storaged...

Reply:
I also check the case for `num_columns:1`, the offset is `Footer offset: 287, size: 164`. Maybe footer is 1/2 - 1/4 size of file in this scenerio

Reply:
"Metadata" is not only the footer, it's all the Thrift overhead + other signaling : page headers, statistics...

Reply:
You're right

```
    auto reader = ParquetFileReader::Open(source, props);
    auto metadata = reader->metadata();
```

Any way, this only reads from footer?

Reply:
I think the concern from @mapleFU is that the write and read paths in this benchmark are not identical. On the read path, only file footer is loaded. However, there are more work done on the write path (e.g. encoding and writing page headers).

Reply:
Ah, I see, that's a good point. I can try to rewrite the write benchmark so as to only measure the metadata write part of writing a file, though that's dependent on where exactly it happens in the Parquet API calls.